src/stan/model/indexing/assign_cl.hpp

```diff
@@ -0,0 +1,276 @@
+  math::arena_matrix_cl<double> prev_vals = lhs;
+  lhs = math::value_of(expr_rhs);  // assign the values
+  math::reverse_pass_callback(
+      [expr_lhs, expr_rhs, name, prev_vals,
+       row_index_tup = std::tuple<RowIndex>(std::forward<RowIndex>(row_index)),
+       col_index_tup
+       = std::tuple<ColIndex>(std::forward<ColIndex>(col_index))]() mutable {
+        decltype(auto) lhs_val
+            = rvalue(expr_lhs.val_op(), name, std::get<0>(row_index_tup),
+                     std::get<0>(col_index_tup));
+        decltype(auto) lhs_adj
+            = rvalue(expr_lhs.adj(), name, std::get<0>(row_index_tup),
+                     std::get<0>(col_index_tup));
+        math::results(lhs_val, math::adjoint_of(expr_rhs), lhs_adj)
+            = math::expressions(
+                prev_vals,
+                math::calc_if<!is_constant<ExprRhs>::value>(
+                    math::adjoint_of(expr_rhs) + lhs_adj),
+                math::constant(0.0, lhs_adj.rows(), lhs_adj.cols()));
+      });
+}
+// the "forwarding" overload
+inline void assign(math::var_value<math::matrix_cl<double>>& expr_lhs,
+                   math::var_value<math::matrix_cl<double>>&& expr_rhs,
+                   const char* /*name*/, index_omni /*row_index*/,
+                   index_omni /*col_index*/) {
+  expr_lhs.vi_ = expr_rhs.vi_;
+}
+
+/**
+ * Assign a primitive or reverse mode scalar to a reverse mode kernel generator
+ * expression, using given uni indices.
+ *
+ * @tparam ExprLhs type of the assignable rev expression on the left hand side
+ * of the assignment
+ * @tparam ScalRhs type of the prim or rev scalar on the right hand side of the
+ * assignment
+ * @param[in,out] expr_lhs expression on the left hand side of the assignment
+ * @param scal_rhs scalar on the right hand side of the assignment
+ * @param name Name of lvalue variable
+ * @param row_index index used for indexing rows of `expr_lhs`
+ * @param col_index index used for indexing columns of `expr_lhs`
+ * @throw std::out_of_range If the index is out of bounds.
+ */
+template <typename ExprLhs, typename ScalRhs,
+          require_rev_kernel_expression_t<ExprLhs>* = nullptr,
+          require_stan_scalar_t<ScalRhs>* = nullptr>
+inline void assign(ExprLhs&& expr_lhs, const ScalRhs& scal_rhs,
+                   const char* name, const index_uni row_index,
+                   const index_uni col_index) {
+  decltype(auto) lhs = math::block_zero_based(
+      expr_lhs.val_op(), row_index.n_ - 1, col_index.n_ - 1, 1, 1);
+  math::arena_matrix_cl<double> prev_val = lhs;
+  lhs = math::constant(math::value_of(scal_rhs), 1, 1);  // assign the value
+  math::reverse_pass_callback(
+      [expr_lhs, scal_rhs, row_index, col_index, prev_val]() mutable {
+        auto&& lhs_val = math::block_zero_based(
+            expr_lhs.val_op(), row_index.n_ - 1, col_index.n_ - 1, 1, 1);
+        auto&& lhs_adj = math::block_zero_based(
+            expr_lhs.adj(), row_index.n_ - 1, col_index.n_ - 1, 1, 1);
```

Comment:
Yeah it probably should. This works as it is just because we dont have a separate overload of `from_matrix_cl` that would actually move rvalue inputs.